cassovary-core/src/main/scala/com/twitter/cassovary/util/collections/CQueue.scala

```diff
@@ -0,0 +1,137 @@
+package com.twitter.cassovary.util.collections
+
+import it.unimi.dsi.fastutil.PriorityQueue
+import it.unimi.dsi.fastutil.ints.{IntArrayFIFOQueue, IntComparator, IntHeapPriorityQueue,
+  IntPriorityQueue}
+import it.unimi.dsi.fastutil.objects.{ObjectArrayFIFOQueue, ObjectArrayPriorityQueue}
+
+/**
+ * Specialized implementation of queue.
+ *
+ * Implemented methods are not boxed, when using with primitive types.
+ */
+abstract class CQueue[@specialized(Int, Long) T] {
+  def first(): T
+  def +=(elem: T)
+  def deque(): T
+  def isEmpty: Boolean
+  def clear(): Unit
+  def size(): Int
+}
+
+/**
+ * FIFO [[CQueue]].
+ */
+trait FIFO[@specialized(Int, Long) T] extends CQueue[T] {
+  def enqueueFirst(elem: T)
```

Comment:
I don't really get what you mean. My thoughts are:
* We instantiate only in the companion object to get implicits working. 
* The only difference I know of between ABC and trait is that trait has no fields, so can be mixed in. Here, trait is ok since FIFO has no fields. 
*It should extend CQueue because FIFO is always a Queue.